Fix memory leak when using OpenSSL and threads

ashman-p opened the pull request:
<!-- Please give a brief explanation of the purpose of this pull request. -->
'run_tests' memory leak tests fails when build options enables OQS_USE_PTHREADS and OQS_USE_OPENSSL.

```
valgrind tests/test_sig SPHINCS+-SHA2-256s-simple
==158054== Memcheck, a memory error detector
==158054== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==158054== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==158054== Command: tests/test_sig SPHINCS+-SHA2-256s-simple
==158054== 
Testing signature algorithms using liboqs version 0.10.2-dev
Configuration info
==================
Target platform:  x86_64-Linux-5.4.0-126-generic
Compiler:         gcc (9.4.0)
Compile options:  [-Wa,--noexecstack;-O3;-fomit-frame-pointer;-fdata-sections;-ffunction-sections;-Wl,--gc-sections;-Wbad-function-cast]
OQS version:      0.10.2-dev
Git commit:       f47817d8cb0cd621dada5276f30f99934f3132a9
OpenSSL enabled:  Yes (OpenSSL 3.3.3.8.3.0-dev )
AES:              NI
SHA-2:            OpenSSL
SHA-3:            OpenSSL
OQS build flags:  OQS_DIST_BUILD OQS_LIBJADE_BUILD OQS_OPT_TARGET=generic CMAKE_BUILD_TYPE=Release 
CPU exts active:  AES AVX AVX2 BMI1 BMI2 PCLMULQDQ POPCNT SSE SSE2 SSE3
================================================================================
Sample computation for signature SPHINCS+-SHA2-256s-simple
================================================================================
verification passes as expected
==158054== 
==158054== HEAP SUMMARY:
==158054==     in use at exit: 240 bytes in 1 blocks
==158054==   total heap usage: 7,654 allocs, 7,653 frees, 844,712 bytes allocated
==158054== 
==158054== LEAK SUMMARY:
**==158054==    definitely lost: 240 bytes in 1 blocks**
==158054==    indirectly lost: 0 bytes in 0 blocks
==158054==      possibly lost: 0 bytes in 0 blocks
==158054==    still reachable: 0 bytes in 0 blocks
==158054==         suppressed: 0 bytes in 0 blocks
==158054== Rerun with --leak-check=full to see details of leaked memory
==158054== 
==158054== For lists of detected and suppressed errors, rerun with: -s
==158054== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

<!-- Does this PR resolve any issue?  If so, please reference it using automatic-closing keywords like "Fixes #123." -->
Fixes #1941.

<!-- Any PR adding a new feature is expected to contain a test; the test should be part of CI testing, preferably within the ".github/workflows" directory tree. Please add an explanation to the PR if/when (why) this cannot be done. -->

<!-- Please answer the following questions to help manage version and changes across projects. -->

* [ ] Does this PR change the input/output behaviour of a cryptographic algorithm (i.e., does it change known answer test values)?  (If so, a version bump will be required from *x.y.z* to *x.(y+1).0*.)
* [ ] Does this PR change the list of algorithms available -- either adding, removing, or renaming? Does this PR otherwise change an API? (If so, PRs in fully supported downstream projects dependent on these, i.e., [oqs-provider](https://github.com/open-quantum-safe/oqs-provider) will also need to be ready for review and merge by the time this is merged.)

<!-- Once your pull request is ready for review and passing continuous integration tests, please convert from a draft PR to a normal PR, and request a review from one of the OQS core team members. -->

ashman-p committed:
init OpenSSL when OQS_USE_OPENSSL is enabled in liboqs.

Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p committed:
wrap OPENSSL_init_crypto with OSSL_FUNC.

Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p committed:
Add OPENSSL_init_crypto to the list of OpenSSL functions called liboqs.

Signed-off-by: Norman Ashley <nashley@cisco.com>

dstebila commented:
This is present in the 0.11.0 release, I guess?  Is it critical to fix?

Do we have a CI job that tests this configuration?

ashman-p commented:
> I'm not able to reproduce the leak myself—@ashman-p could you let me know what Linux environment you're running so I can test various configs out in Docker images?

@SWilson4 , Thanks for checking on this issue.
I just have a generic linux setup.
Linux nashley-dev 5.4.0-126-generic #142-Ubuntu SMP Fri Aug 26 12:12:57 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux


liboqs Build

cmake -DBUILD_SHARED_LIBS=ON -DOQS_USE_OPENSSL=ON -DCMAKE_BUILD_TYPE=Debug -GNinja ..
-- The C compiler identification is GNU 9.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Performing Test CC_SUPPORTS_WA_NOEXECSTACK
-- Performing Test CC_SUPPORTS_WA_NOEXECSTACK - Success
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Algorithms filtered for KEM_ml_kem_512;KEM_ml_kem_768;KEM_ml_kem_1024;KEM_ml_kem_512_ipd;KEM_ml_kem_768_ipd;KEM_ml_kem_1024_ipd;SIG_ml_dsa_44;SIG_ml_dsa_65;SIG_ml_dsa_87;SIG_sphincs_sha2_128f_simple;SIG_sphincs_sha2_128s_simple;SIG_sphincs_sha2_256f_simple;SIG_sphincs_sha2_256s_simple;SIG_sphincs_shake_128f_simple;SIG_sphincs_shake_128s_simple;SIG_sphincs_shake_256f_simple
-- Found OpenSSL: /home/ubuntu/openssl3.3/ossl-install/lib/libcrypto.so (Required is at least version "1.1.1")  
-- Looking for aligned_alloc
-- Looking for aligned_alloc - found
-- Looking for posix_memalign
-- Looking for posix_memalign - found
-- Looking for memalign
-- Looking for memalign - found
-- Looking for explicit_bzero
-- Looking for explicit_bzero - found
-- Looking for explicit_memset
-- Looking for explicit_memset - not found
-- Looking for memset_s
-- Looking for memset_s - not found
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.17") found components: doxygen dot 
-- Configuring done
-- Generating done

steenrasmussen committed:
Don't include dlfcn.h for Windows (#1936)

Signed-off-by: Steen Rasmussen <steen.rasmussen@dencrypt.dk>
Co-authored-by: Steen Rasmussen <steen.rasmussen@dencrypt.dk>
Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p commented:
@SWilson4, its the use of OQS_USE_SHA3_OPENSSL=ON  that causes the problem.

SWilson4 commented:
OK, I've managed to reproduce the leak, but only when building against OpenSSL >= 3.3.2. In particular, the leak does _not_ occur when building against OpenSSL 3.3.1 with the same configuration.

It seems to me that this might actually be an OpenSSL bug rather than a liboqs bug, especially as the fix proposed here should (?) be unnecessary: per the [OpenSSL docs](https://docs.openssl.org/3.0/man3/OPENSSL_init_crypto):
> As of version 1.1.0 OpenSSL will automatically allocate all resources that it needs so no explicit initialisation is required. Similarly it will also automatically deinitialise as required.

Looping in @baentsch @romen @beldmit @levitte: any knowledge of a regression going from 3.3.1 to 3.3.2?

I will try to isolate the exact commit which introduces the leak.

beldmit commented:
At first glance nothing suspicious

SWilson4 commented:
> I will try to isolate the exact commit which introduces the leak.

It seems that this leak was introduced in https://github.com/openssl/openssl/commit/83efcfdfa1de760bd30df7f4cf94e7a0d2b0db9f. When building against the parent of that commit (https://github.com/openssl/openssl/commit/a13df68796828794920403c31d77409b0f06bae0) the leak does not occur.

SWilson4 commented:
^ fyi @beldmit

beldmit commented:
@nhorman could you please take a look?

nhorman commented:
FWIW, i don't see anything expressly wrong with the changes, its fine to call OPENSSL_init_crypto in your application rather than having the library do it as needed, but I'm having a hard time understanding:
1) Where exactly the leak is
2) How manually calling OpenSSL_init_crypto fixes it

I'll try reproduce here, but since you already have it set up, can you re-run valgrind with --leak-check=full to show the stack trace of where the leaked memory is being allocated?

nhorman commented:
Thank you.  Thats interesting, I just tried to recreate with the head of the master branch, and didn't encounter the leak.  Given that you reported the issue was introduced with commit 83efcfd (which is a backport of the original commit to the 3.3 branch, that makes me think we fixed something in master that we didn't backport.  Let me see if I can find it

SWilson4 commented:
> Thank you. Thats interesting, I just tried to recreate with the head of the master branch, and didn't encounter the leak. Given that you reported the issue was introduced with commit 83efcfd (which is a backport of the original commit to the 3.3 branch, that makes me think we fixed something in master that we didn't backport. Let me see if I can find it

Weird—I'm also getting the leak on the latest master (https://github.com/openssl/openssl/commit/c262cc0c0444f617387adac3ed4cad9f05f9c526).

nhorman commented:
hmm, what version of valgrind are you using?  I'm on valgrind-3.23.0

nhorman commented:
In fact, I'm sure its my config, I just ran it through gdb and never hit a breakpoint in ossl_rcu_read_lock

nhorman commented:
There we go, if I load the oqs provider (duh, should have thought of that), I see the leak.  Now to figure out why

SWilson4 commented:
I'm on valgrind-3.22.0, and I haven't made any manual changes to openssl.cnf (including to load any provider).

I'm running everything in a Docker container on x86_64 / Linux; I'll attach a Dockerfile to duplicate my setup in case it's helpful to you.

SWilson4 commented:
```dockerfile
FROM openquantumsafe/ci-ubuntu-latest:latest

WORKDIR /root

RUN git clone --depth=1 --branch=0.11.0 https://github.com/open-quantum-safe/liboqs.git liboqs
RUN git clone --branch=master https://github.com/openssl/openssl.git openssl

WORKDIR /root/openssl

RUN ./config --debug --prefix=/root/.localopenssl && make -j && make install_sw install_ssldirs

WORKDIR /root/liboqs

RUN mkdir build

WORKDIR ./build

RUN cmake -GNinja -DOPENSSL_ROOT_DIR=../.localopenssl -DOQS_MINIMAL_BUILD="SIG_ml_dsa_65" -DOQS_USE_SHA3_OPENSSL=ON -DCMAKE_BUILD_TYPE=Debug ..
RUN ninja
```

nhorman commented:
Ok, think I found the problem. https://github.com/openssl/openssl/commit/83efcfdfa1de760bd30df7f4cf94e7a0d2b0db9f added a use of ossl_init_thread_start to the code base, for which there is only a handful of users.  ossl_init_thread_start registers a handler function to clean up thread local data, not when the thread exits explicitly, but rather when the library context against which the thread allocates resources is deleted.  I'm not 100% sure why we do this, but its how it works.  

Normally its not a problem.  Library contexts get deleted in OSSL_LIB_CTX_free, that calls context_deinit->ossl_ctx_thread_stop->init_thread_stop, and that calls all the handlers registered via ossl_init_thread_start, and everything gets cleaned up.

However, in liboqs, you implicitly call the OpenSSL_cleanup routine via the registered atexit() handler, which again, is ok, but, it means that in test_sig.c (and likely your other tests), you join and exit the thread prior to OpenSSL getting cleaned up.  It appears whats happening is that, on exit, the C library is cleaning up its thread local data list (I feel like I investigated this before), and, not having an cleanup handler registered (recalling from above that we register the local data handler with libcrypto, _not_ with libc), and so it just NULL's the pointer in the thread local data table in the c library.  As a result when OpenSSL_cleanup is called (either explicitly or implicitly), we run through the libcrypto registered cleanup handlers, the handler for rcu gets called (ossl_rcu_free_local_data), and when we call CRYPTO_THREAD_get_local to fetch the local data to free (which maps to pthread_getspecific) it returns NULL, as libc already expunged that pointer.  As  a result the local data leaks.

This API could use some reworking, as It seems to me that it relies on implementation details in libc that aren't guaranteed (I think in the past those pointers stuck around until after all the exit handlers ran, but I need to look further).  The documentation here:
https://docs.openssl.org/3.0/man3/OPENSSL_init_crypto/#description
Is sort of vague on the subject, indicating that the cleanup should be done on thread exit, but thats not really true, as its actually done on behalf of the thread when the library context is cleaned.

For the time being, as any fix here would be an ABI break, I think its something we need to live with.  The good news(?) is that theres an API to manage this, and theres precedent for it in our tests.  OPENSSL_thread_stop is used to explicitly clean up a threads local data prior to exiting, which can be used here.  You can see it used in situations similar/identical  to yours in drbgtest.c and threadstest.h in which thread wrappers like thread_run() call it prior to exiting for just this reason.

You will likely need to do this for your various thread enabled tests, but this patch:

```
diff --git a/tests/test_sig.c b/tests/test_sig.c
index e94d3034..bd5e4728 100644
--- a/tests/test_sig.c
+++ b/tests/test_sig.c
@@ -9,6 +9,7 @@
 #include <string.h>
 
 #include <oqs/oqs.h>
+#include <openssl/crypto.h>
 
 #if OQS_USE_PTHREADS
 #include <pthread.h>
@@ -183,6 +184,7 @@ struct thread_data {
 void *test_wrapper(void *arg) {
        struct thread_data *td = arg;
        td->rc = sig_test_correctness(td->alg_name);
+       OPENSSL_thread_stop();
        return NULL;
 }
 #endif
```

which I tested here, resolves the leak for me

ashman-p commented:
Hi @nhorman, thanks for tracking this down. An observation I made earlier when this issue was discovered was that it seems to manifest with SHA3 fetches but not SHA2. Does this still fit with your findings?

nhorman commented:
It's a bit late here now, but I'll take a look in the am and let you know

nhorman commented:
@ashman-p do you have a particular signature algorithm that you see as not encountering the leak?  I'm looking at the available defined SIG algs and I don't see any available that make use of SHA3

baentsch commented:
> resolves the leak for me

Thanks @nhorman for the analysis and the proposed fix. I guess the proper location for your fix would be [here](https://github.com/open-quantum-safe/liboqs/blob/7f4c89b26fc2a59dd314950f1663eaff037dff72/src/common/ossl_helpers.c#L76-L88)?

baentsch commented on src/common/common.c:
```diff
@@ -228,6 +228,10 @@ OQS_API int OQS_CPU_has_extension(OQS_CPU_EXT ext) {
 OQS_API void OQS_init(void) {
+#if defined(OQS_USE_OPENSSL)
+	OSSL_FUNC(OPENSSL_init_crypto)(OPENSSL_INIT_ENGINE_ALL_BUILTIN | OPENSSL_INIT_ADD_ALL_CIPHERS,
+	                               NULL);
```
Thanks for the proposal @ashman-p -- but allow me to ask why an additional init call would fix some apparently missing memory cleanup (at exit)? Or can you explain the choice of parameters here that would facilitate that? From the [documentation of those](https://docs.openssl.org/1.1.1/man3/OPENSSL_init_crypto/#description) it's not quite obvious to me (why these have been chosen and how they contribute to fixing the issue):

> OPENSSL_INIT_ENGINE_ALL_BUILTIN

> With this option the library will automatically load and initialise all the built in engines listed above with the exception of the openssl and afalg engines. This not a default option.

> OPENSSL_INIT_ADD_ALL_CIPHERS

> With this option the library will automatically load and make available all libcrypto ciphers. This option is a default option. Once selected subsequent calls to OPENSSL_init_crypto() with the option OPENSSL_INIT_NO_ADD_ALL_CIPHERS will be ignored.

ashman-p replied:
@baentsch, memory leak test failures occurred in test_kem, test_signatures, ... these options seemed to address all the failures. Neil's and Spencer's findings are perhaps pointing more to a root cause.

baentsch replied:
> are perhaps pointing more to a root cause

And (finding and solving) that would be my preference. A workaround imo carries the risk of "getting bitten" in the future again (and/or reduces the opportunity to improve any OSS, be the problem in OQS or OpenSSL). Time permitting of course. Apologies for being anal, @ashman-p . Closing the comment such as to not block a merge.

nhorman commented:
@baentsch yes, that seems like a reasonable approach I think.

baentsch commented:
@SWilson4 @ashman-p @dstebila Please also take note of this guidance from [OpenSSL documentation](https://docs.openssl.org/1.1.1/man3/OPENSSL_init_crypto/#notes):

> Similarly this message will also not be sent if OpenSSL is linked statically, and therefore applications using static linking should also call OPENSSL_thread_stop() on each thread. 

and in addition

> On Linux/Unix where OpenSSL has been loaded via dlopen() and the application is multi-threaded and if dlclose() is subsequently called prior to the threads being destroyed then OpenSSL will not be able to deallocate resources associated with those threads. The application should either call OPENSSL_thread_stop() on each thread prior to the dlclose() call, or alternatively the original dlopen() call should use the RTLD_NODELETE flag (where available on the platform).

So I think it is actually _mandatory_ for `liboqs` to call this function given it should work both statically linked as well as when used via dlopen() -- which I think is the primary use case (and surely when used as part of (a dynamically loaded) `oqsprovider`), no? Only question is how to ensure it's called _on each thread_?

The more I read the more I wonder how/why this ever worked OK given this guidance and by-default use of `pthreads`. The sig test also is not the only one _actually using_ threading: At first glance, KEM and SIG_STFL testing do the same...

ashman-p commented:
> ```shell
> cmake -GNinja -DOPENSSL_ROOT_DIR=../.localopenssl_83efcfdfa1_debug -DOQS_MINIMAL_BUILD="SIG_ml_dsa_65" -DOQS_USE_SHA3_OPENSSL=ON -DCMAKE_BUILD_TYPE=Debug ..
> ```
@nhorman, I meant to say that the problem only manifested with "-DOQS_USE_SHA3_OPENSSL=ON". Not that there is a sig test actually uses sha3. Defining OQS_USE_OPENSSL=ON means the OQS library uses OpenSSL's SHA2. My config additionally sets OQS_USE_SHA3_OPENSSL=ON. Without this, Spencer could not recreate the problem. Thats why I asked the question about differences between these 2 options. It seems to me that the problem would be present regardless.
 fetch_ossl_objects(void) in common/ossl_helpers.c calls EVP_MD_fetch() for sha2 and sha3. But, I am not sure where sha3 is used.

bhess committed:
Update CBOM format to upstream v1.6 (#1834)

Signed-off-by: Basil Hess <bhe@zurich.ibm.com>
Signed-off-by: Norman Ashley <nashley@cisco.com>

nhorman commented:
@ashman-p yes, I would think so, I'm a bit tied up at the moment, but given what we've found I would imagine that the leak would occur in that configuration as well, and be solved by the same proposed fix

baentsch commented:
> fetch_ossl_objects(void) in common/ossl_helpers.c calls EVP_MD_fetch() for sha2 and sha3. But, I am not sure where sha3 is used.

That is a very good observation, @ashman-p -- [this code](https://github.com/open-quantum-safe/liboqs/blob/329869f5592e555923eedd100249d2a10f546f3e/src/common/ossl_helpers.c#L25-L73) now seems dubious to me: Shouldn't the EVP-alg-fetchers be contingent/ifdef'd upon the [3 corresponding build options](https://github.com/open-quantum-safe/liboqs/blob/main/CONFIGURE.md#oqs_use_openssl)? What's your take on that @beldmit ?

beldmit commented:
@baentsch you are 100% correct, they should be properly wrapped by `#ifdef`s - if we don't use SHA3 or AES from OpenSSL, it makes no sense to fetch it.

I have originally written this prefetching code because of performance reasons instead of implicit fetching used before, but AFAIK there were many changes since then - e.g. an option to provide 3rd-party callbacks etc so this should be reconsidered

ashman-p commented:
> @ashman-p yes, I would think so, I'm a bit tied up at the moment, but given what we've found I would imagine that the leak would occur in that configuration as well, and be solved by the same proposed fix

No worries Neil. Again, valgrind only reports a leak with the SHA3 macro enabled. Since we now have 2 ways to make the leak go away (your findings and my cheap init crypto call) it makes me wonder if there is something more that could turn up down the road. if you have an idea that i can help to track down (without too much hand holding), I would be happy to try. The test fails both with KEMS as well as signatures.

ashman-p commented:
> There we go, if I load the oqs provider (duh, should have thought of that), I see the leak. Now to figure out why

@nhorman, now getting some time to revisit this today. I missed the significance of your statement here.
Could we be tracking two different issues? My test setup does not involve oqsprovider. I just have openssl 3.3 and liboqs. Run the test and ...

Anyhow, I will apply your patch and report back.

nhorman commented:
@ashman-p I'm not sure, but I'm sufficiently unfamiliar with oqs to fully understand how you would use both together without making use of the liboqs provider.  It's certainly possible that there are two issues here, but the fact that the errors.are identical makes me biased towards thinking the problems are simmilar if not identical

baentsch commented:
> sufficiently unfamiliar with oqs to fully understand how you would use both together without making use of the liboqs provider

Thanks again for taking the time to look into this @nhorman . `liboqs` is a crypto library making available for external consumption via a common API various PQ algorithms; some of these PQ algorithms need "common" things like PRNG and SHA3. `liboqs` therefore has a build option to utilize these common things from OpenSSL's `libcrypto`. That is what @ashman-p has been testing. `liboqs` also can be built with "its own" (really, from other upstream sources) common code, offering users an option to build `liboqs` without OpenSSL dependencies. In that configuration, the problems in this issue don't appear, hinting at least at a wrong interaction pattern between `liboqs` and `libcrypto`.

The `oqsprovider` in turn is a shared library pretty much devoid of crypto functionality (except when implementing hybrid PQ) primarily making available PQ crypto implemented in `liboqs` to the OpenSSL EVP API via the `openssl` provider API.

```

|------------------------------|
|     libssl, openssl apps     |
|------------------------------|
|          EVP API             |
|------------------------------|
|         oqsprovider          |
|------------------------------|

|------------------------------|
|           liboqs             |
|------------------------------|
| PQ alg 1 || .... || PQ alg n |
|------------------------------|
|    openssl    ||   external  |
|   libcrypto   ||  sha3/prng  |
|------------------------------|
```

nhorman commented:
ok, thank you for that.

thats interesting, so liboqs is one of those interesting libraries for which applications may use the provider interface to access oqs, and liboqs in turn loads symbols from openssl to do its work.

The implication here is that we may have two dependencies on openssl (listed via DT_NEEDED entries in the app and the liboqs provider), and those instances may point to different versions of libcrypto.   That gets tricky.

to make this a bit more concrete, and pull it back to the issue at hand, what config file does the liboqs instance of libcrypto load?  Can you post it here?  Dependent on how its built and initalized, it may be that the liboqs instance of libcrypto is loading a "vestigual" provider in this test environment, which is never used, but still triggers the allocation that we are failing to free (which would still point to the thread_stop solution we  have above as the 'correct' solution to the problem.

baentsch commented:
> those instances may point to different versions of libcrypto. That gets tricky.

That absolutely is a problem (known to anyone integrating the whole stack and possibly utilizing distros -- which is not many people :)

But it is not at play here as the issue occurs in "standalone" `liboqs`; no `oqsprovider` involved.

The config (build) and config file question for a reproducer I cannot answer: @ashman-p @SWilson4 ?

nhorman commented:
@baentsch thats not entirely true, and was the point I was getting at.  Even if the top level application doesn't use openssl, and the oqsprovider to access the liboqs library, there is still a path to load the provider.  If:
1) The libcrypto.so library gets loaded by liboqs (as your diagram illustrates)
2) The libcrypto.so library is _not_ initalized with the OPENSSL_INIT_NO_LOAD_CONFIG flag (likely if you use auto initalization)
3) The default config file for openssl (or whatever file OPENSSL_CONF points to) contains a section pointing to the oqs provider

Then, when all of those conditions are true, libcrypto will still load the oqs provider, even though the library doesn't use it, which will send you down the path [here](https://github.com/open-quantum-safe/liboqs/pull/1942#issuecomment-2389538634), and you'll get the leak

So in short, if you have a config file that references the oqs provider (or any external provider, like fips), then this issue will be the result, even if the application being run doesn't actually make use of it

baentsch commented:
> there is still a path to load the provider.

Fully agree, @nhorman . It was my _strong_ assumption that such circular setup has been excluded in testing as both @ashman-p and @SWilson4 know about the dependencies (and https://github.com/open-quantum-safe/liboqs/pull/1942#issuecomment-2389538634 seems to show a standalone build -- and https://github.com/open-quantum-safe/liboqs/pull/1942#issuecomment-2389711126 absolutely guarantees one).

But you're right in checking this explicitly by asking for the config -- particularly considering the problem never materialized in CI. Also helpful may simply be the output of `openssl list -providers` in a failing setup.

ashman-p commented:
> @ashman-p can you identify the openssl.cnf file that you are using for the libcrypto instance that liboqs uses and post it here? That would clarify things greatly.

In my setup it was the vanilla openssl.cnf with the default provider.
apps/openssl list -providers
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.3.3
    status: active

nhorman commented:
That would do it.  the default provider gets loaded via:
module_run->do_load_builtin_modules->OPENSSL_load_builtin_modules->ossl_provider_add_conf_module->CONF_module_add->module_add

which then allocates the lock in question thats leaking

I avoided that initially as I didn't have the default provider activated in my config.  Once I loaded the oqs provider (which could have been any provider, it was just the one I chose, as I was working with oqs), the problem began to appear

SWilson4 commented:
Here's the configuration in my environment. (Added the .txt extension for GitHub upload only.)

[openssl.cnf.txt](https://github.com/user-attachments/files/17259441/openssl.cnf.txt)

nhorman commented:
yeah, that looks to be the same situation

ashman-p commented:
Ok, thank you @nhorman. So, my next question is ... base my tests with "OPENSSL_thread_stop()".
This only worked when called, as Neil originally prescribed. i.e from the thread-based application. Calling it from any centrally used function like oqs_ossl_destroy() did work. Presumable because the thread is already suspended or ended by that time?

Question: We have 2 work-around options... OPENSSL_thread_stop() and OPENSSL_init_crypto().
Is there any reason not to go with the 'OPENSSL_init_crypto' option? It is simpler to implement rather than changing all the tests apps. Also, this path means we would also need to instruct application writers to also make this call.

baentsch commented:
> Calling it from any centrally used function like oqs_ossl_destroy() did work. 

I suppose this should read "...did NOT work.", right @ashman-p ?

baentsch commented:
> It looks like the leak does indeed occur for SHA2 as well. 

And does it disappear likewise if `OPENSSL_thread_stop` is added to the corresponding test, @SWilson4 ?

ashman-p commented:
> > Calling it from any centrally used function like oqs_ossl_destroy() did work.
> 
> I suppose this should read "...did NOT work.", right @ashman-p ?

That is correct. It did **NOT** work. Sorry for the omission.

ashman-p commented:
> > Is there any reason not to go with the 'OPENSSL_init_crypto' option?
> 
> It's easier indeed -- but it'd really be good to understand _why_ it is working. Without this understanding this may haunt us in the future again, say if some side effect in `openssl` initialization making it work right now changes....
> 
> The "thread_stop" solution is clearly understandable and documented, would require 3 code changes in the test programs and one documentation change pointing to OpenSSLs guidance: Would that be too undesirable, @ashman-p ?

Yeah, i hear you. I am fine with us doing that. I wondered is @nhorman had any idea why the init calls seem to affect things? The other concern i have is for pre-existing applications. Seems possible that there would be some fallout?

SWilson4 commented:
> The "thread_stop" solution is clearly understandable and documented, would require 3 code changes in the test programs and one documentation change pointing to OpenSSLs guidance: Would that be too undesirable, @ashman-p ?

My preference would be to go with the "thread_stop" solution, for the reasons above. Also, if the leak only occurs because of a flaw in our test programs and _not_ because of a flaw in the library itself, then that's great news, and I would endorse patching it only there.

I do wonder, though, if we ought to protect the user a little more—it seems wrong somehow that somebody who wants to use liboqs in a multithreaded application is required to
1. be aware that we're using OpenSSL under the hood (what if they get liboqs via their package manager and are unaware of the build command used?)
2. call the appropriate OpenSSL (_not_ liboqs) cleanup function in order to avoid a memory leak.

The only way I see to avoid this pain for the user is to add an analogous well-documented `OQS_thread_stop` function—this could be implemented using our usual #ifdef OQS_USE_OPENSSL approach. Thoughts on this?

> > It looks like the leak does indeed occur for SHA2 as well.
> 
> And does it disappear likewise if `OPENSSL_thread_stop` is added to the corresponding test, @SWilson4 ?

Yes, it does.

baentsch commented:
> The only way I see to avoid this pain for the user is to add an analogous well-documented OQS_thread_stop function—this could be implemented using our usual #ifdef OQS_USE_OPENSSL approach. Thoughts on this?

Sounds (more) sensible (than pointing to openssl) -- but doesn't alleviate users of the need to call this at the right time.

> Yes, it does.

Thanks for checking/confirming, @SWilson4 !

> The other concern i have is for pre-existing applications. Seems possible that there would be some fallout?

I don't see which: The problem also is "pre-existing" -- so either it didn't bother folks (so a fix with a new API they call (or don't :) doesn't change anything) or they already properly cleaned up "behind themselves" as per OpenSSL guidance and only the OQS tests didn't do so far (again, not making a difference for other pre-existing apps).

ashman-p commented:
@SWilson4, I can complete the code changes if you haven't already done them. But can you work out the documentation?

SWilson4 commented:
> @SWilson4, I can complete the code changes if you haven't already done them. But can you work out the documentation?

Sounds like a plan, thanks @ashman-p.

ashman-p committed:
Revert files

Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p committed:
Cleanup threading resources when OpenSSL is used.

Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p committed:
Add OPENSSL_thread_stop to the ossl function list. Cleanup error checking.

Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p committed:
Update ossl_helpers.c

Remove OPENSSL build-time macro from helper functions.

Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p commented:
@baentsch, @nhorman, can you give me any pointers on resolving the errors with 'dlopen' builds? Undefined ref to OPENSSL_thread_stop when -DOQS_DLOPEN_OPENSSL=ON is used?
Thanks.

baentsch commented:
> @baentsch, @nhorman, can you give me any pointers on resolving the errors with 'dlopen' builds? Undefined ref to OPENSSL_thread_stop when -DOQS_DLOPEN_OPENSSL=ON is used? Thanks.

Tagging @ueno as the author of that option.

ueno commented:
> > @baentsch, @nhorman, can you give me any pointers on resolving the errors with 'dlopen' builds? Undefined ref to OPENSSL_thread_stop when -DOQS_DLOPEN_OPENSSL=ON is used? Thanks.

With `dlopen` builds, OpenSSL functions are supposed to be called in a form `OSSL_FUNC(function)(args)`, after including `src/common/ossl_helpers.h`. The problem, however, is that the header file is not exposed to the tests. My suggestion (haven't tried it yet) is to create a thin wrapper around `OPENSSL_thread_stop`, say `OQS_thread_stop` in `src/common/common.c`, and expose it as an internal API (i.e., without `OQS_API`).

baentsch commented:
> say OQS_thread_stop in src/common/common.c, expose it as an internal API [...] and use it throughout the tests

OK -- that might solve the issue for "internal use" (tests), but isn't the proposal to expose `OQS_thread_stop` externally ([see above](https://github.com/open-quantum-safe/liboqs/pull/1942#issuecomment-2393958431))?

ashman-p commented on tests/test_sig_stfl.c:
```diff
@@ -1006,19 +1008,28 @@ void *test_create_keys(void *arg) {
 	struct lock_test_data *td = arg;
 	printf("\n%s: Start Generate Keys\n", __func__);
 	td->rc = sig_stfl_test_secret_key_lock(td->alg_name, td->katfile);
+#if defined(OQS_USE_OPENSSL)
```
```suggestion
#if defined(OQS_USE_OPENSSL) && !defined(OQS_DLOPEN_OPENSSL)
```

ashman-p commented on tests/test_sig.c:
```diff
@@ -183,6 +186,9 @@ struct thread_data {
 void *test_wrapper(void *arg) {
 	struct thread_data *td = arg;
 	td->rc = sig_test_correctness(td->alg_name);
+#if defined(OQS_USE_OPENSSL)
```
```suggestion
#if defined(OQS_USE_OPENSSL) && !defined(OQS_DLOPEN_OPENSSL)
```

ashman-p commented on tests/test_kem.c:
```diff
@@ -206,6 +209,9 @@ struct thread_data {
 void *test_wrapper(void *arg) {
 	struct thread_data *td = arg;
 	td->rc = kem_test_correctness(td->alg_name);
+#if defined(OQS_USE_OPENSSL)
```
```suggestion
#if defined(OQS_USE_OPENSSL) && !defined(OQS_DLOPEN_OPENSSL)
```

ashman-p committed:
Apply suggestions from code review

Use of OQS_DLOPEN_OPENSSL needs a new API change to expose an OQS function that calls  OPENSSL_thread_stop

Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p committed:
Update tests/test_sig.c

Use of OQS_DLOPEN_OPENSSL needs a new API change to expose an OQS function that calls  OPENSSL_thread_stop

Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p committed:
Update tests/test_kem.c

Use of OQS_DLOPEN_OPENSSL needs a new API change to expose an OQS function that calls  OPENSSL_thread_stop

Signed-off-by: Norman Ashley <nashley@cisco.com>

ashman-p commented:
This last commit isolates the memory leak fix when dlopen is used. A separate will be created to track the addition of a new OQS API needed to address this. I hope this is a reasonable and agreeable approach as discussed earlier in a status meeting.

SWilson4 commented:
It looks like some unrelated changes to CBOM-related files snuck into the diff. Possibly a rebase gone wrong @ashman-p?

ashman-p commented:
> It looks like some unrelated changes to CBOM-related files snuck into the diff. Possibly a rebase gone wrong @ashman-p?

Yeah, sorry about that. I am going to punt this PR and start over. New PR is
[here.](https://github.com/open-quantum-safe/liboqs/pull/1959)

ashman-p commented:
Starting over...
